hashing password

carvilsi opened the pull request:
The passwords stored at lokijs will be hashed using [bcrypt](https://github.com/kelektiv/node.bcrypt.js)

gvdent commented:
Changes look great 👍 

There is two things however: 
- This code will need to work with existing databases of users (it currently works for new users, not existing ones). The way I see it there is two options: Either bcrypt the stored plain-text password on first (attempt to) login of the user, or run a onetime updatescript on the loki stored data on startup of weaver-server. I prefer the latter.
- The admin password is currently stored in plain text in the config file, this will also need to be configured there through a hash (and with that, a small readme of how to actually generate that hash for your friendly system administator ;)).

carvilsi commented:
- I agree with run onetime updateScript.
- About the admin password, my idea was solving it in other PR. But also makes sense to resolve it in this one. (and be sure that my friendly system administrator will have this readme ;)

carvilsi commented:
- Added hashing for admin user on config/default.coffee
- Removed admin password from splash screen
- Created init script to rewrite old plain text passwords to the hashed ones
- Created hash script to create new weaver admin user (for my friendly system administator ;))

gvdent commented on src/auth/UserService.coffee:
```diff
@@ -51,7 +51,7 @@ class UserService extends LokiService
       throw {code: -1, message: "User not found"}
```
Minor: The message should be the same for invalid passwords as for users not found, to prevent information about the account existing from leaking.

gvdent commented on config/default.coffee:
```diff
@@ -1,12 +1,13 @@
   admin:
     username: 'admin'
-    password: 'admin'
+    password: '$2a$10$CnY1NFwHHo1v0qiF/rKgIOWAxUOA5Znh7kMPoh.Ru98uX9CZ7MPqC'
```
For easier installation/experimentation on development: Could you introduce a seperate config item for the hashedPassword, and then include logic to use the hashedPassword if that is available and a regular plain text password otherwise? (if both are provided prefer the hashed version).

carvilsi replied:
I'm not sure… For installation or experimentation on development there are two procedures:
- the hashed password on development branch is 'admin'
- If you need to modify this default password, you must execute the script to generate an admin hashed password.

I think that complicating this, allowing to coexist the passwords in plain text and hashed version, we are increasing the chances to use the password in plain text instead the hashed one. Therefore reducing the security.
In my opinion, the admin password must be only written in hashed version.

gvdent replied:
Actually I agree... Ok, I'll fix up something for installation at client sites then.

carvilsi replied:
Perfect! Please, let me know if some help is need it with it.
I think that the PR is ready to merge (after reviewing)

gvdent commented on src/auth/AdminUser.coffee:
```diff
@@ -17,7 +18,10 @@ class AdminUser
 
   signInUsername: (username, password) ->
-    username is @username and password is adminPass
+    bcrypt.compare(password,adminPass)
```
Yay, no timing attacks :)

gvdent commented:
This PR does break automatic login on development machines in the browser. Since this is development convenience functionality which could be considered a security risk anyway, breaking this is fine.
